File: #221_Maximal_Square/Solution.java
/* O(m * n) solution */
public class Solution {
    public int maximalSquare(char[][] matrix) {
        if (matrix == null || matrix.length == 0) return 0;
        int rows = matrix.length;
        int cols = matrix[0].length;
        int[][] dp = new int[rows][cols];
        int max = 0;
        for (int i = 1; i < rows; i++) {
        	for (int j = 1; j < cols; j++) {
        		if (matrix[i][j] == '0') continue;
        		if (i == 0 || j == 0) dp[i][j] = 1;
        		else {
        			int min = Math.min(dp[i - 1][j - 1], Math.min(dp[i][j - 1], dp[i - 1][j]));
        			int sqrt_min = (int) Math.sqrt(min);
	        		dp[i][j] = (sqrt_min + 1) * (sqrt_min + 1);
	        	}
        		max = Math.max(dp[i][j], max);
        	}
        }
        return max;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	char[][] t = {
    		{'1', '1', '1', '1'},
			{'1', '1', '1', '1'},
			{'1', '1', '1', '1'}
    	};
    	System.out.println(sol.maximalSquare(t));
    }
}

File: #221_Maximal_Square/Solution.py
class Solution(object):
    def maximalSquare(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        if not matrix or not matrix[0]:
        	return 0
        maxSide = 0
        dp = [[0 for j in range(len(matrix[0]))] for i in range(len(matrix))]
        for j in range(len(matrix[0])):
        	if matrix[0][j] == '1':
        		dp[0][j] = 1
        		maxSide = max(maxSide, dp[0][j])
        for i in range(len(matrix)):
        	if matrix[i][0] == '1':
        		dp[i][0] = 1
        		maxSide = max(maxSide, dp[i][0])
        for i in range(1, len(matrix)):
        	for j in range(1, len(matrix[0])):
        		if matrix[i][j] == '0':
        			continue
        		dp[i][j] = min(dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1]) + 1
        		maxSide = max(maxSide, dp[i][j])
        return maxSide * maxSide

sol = Solution()
t = [['0', '1'], 
	['1', '0']]
print sol.maximalSquare(t)